src/spec.c

```diff
@@ -1112,206 +1187,339 @@ static void IndexStats_RdbSave(RedisModuleIO *rdb, IndexStats *stats) {
-  //  Document doc = {0};
-  //  Document_Init(&doc, keyname, 1.0, RS_LANG_ENGLISH);
-  //  if (Document_LoadAllFields(&doc, ctx) != REDISMODULE_OK) {
-  //    RedisModule_Log(ctx, "warning", "Failed loading document %*.s", (int)keynameCStrLen,
-  //                    keynameCStr);
-  //    Document_Free(&doc);
-  //    return;
-  //  }
-  //
-  //  QueryError status = {0};
-  //
-  //  dictIterator *iter = dictGetIterator(specDict);
-  //  dictEntry *entry = NULL;
-  //  while ((entry = dictNext(iter))) {
-  //    IndexSpec *sp = dictGetVal(entry);
-  //    if (IndexSpec_ShouldIndex(sp, keyname, key)) {
-  //
-  //      RedisSearchCtx sctx = SEARCH_CTX_STATIC(ctx, sp);
-  //
-  //      RSAddDocumentCtx *aCtx = NewAddDocumentCtx(sp, &doc, &status);
-  //      if (aCtx == NULL) {
-  //        // todo: handle this error
-  //        continue;
-  //      }
-  //
-  //      aCtx->donecb = IndexSpec_DoneIndexingCallabck;
-  //
-  //      aCtx->stateFlags |= ACTX_F_NOBLOCK;
-  //      aCtx->stateFlags |= ACTX_F_NOFREEDOC;
-  //
-  //      AddDocumentCtx_Submit(aCtx, &sctx, DOCUMENT_ADD_REPLACE);
-  //    }
-  //  }
-  //  dictReleaseIterator(iter);
-  //
-  //  // doc was set DEAD in Document_Moved and was not freed since it set as NOFREEDOC
-  //  doc.flags &= ~DOCUMENT_F_DEAD;
-  //  Document_Free(&doc);
+  IndexSpec *sp = scanner->spec_opt;
+  if (sp) {
+    IndexSpec_UpdateMatchingWithSchemaRules(sp, ctx, keyname);
+  } else {
+    Indexes_UpdateMatchingWithSchemaRules(ctx, keyname);
+  }
+  ++scanner->scannedKeys;
 }
 
-void IndexSpec_ScanAndReindexSpec(void *notused) {
+//---------------------------------------------------------------------------------------------
+
+#define REINDEX_BATCH_SIZE 100
+
+static void Indexes_ScanAndReindexAsyncTask(void *_scanner) {
+  RS_LOG_ASSERT(_scanner, "invalid IndexesScanner");
+
+  IndexesScanner *scanner = _scanner;
+  IndexSpec *sp = scanner->spec_opt; // can be NULL
+
   RedisModuleCtx *ctx = RedisModule_GetThreadSafeContext(NULL);
   RedisModuleScanCursor *cursor = RedisModule_ScanCursorCreate();
-  //  RedisModule_ThreadSafeContextLock(ctx);
-  while (RedisModule_Scan(ctx, cursor, IndexSpec_ScanCallback, NULL)) {
-    //    RedisModule_ThreadSafeContextUnlock(ctx);
-    //    RedisModule_ThreadSafeContextLock(ctx);
+  RedisModule_ThreadSafeContextLock(ctx);
+
+  if (sp) {
+    RedisModule_Log(ctx, "notice", "Scanning index %s in background", sp->name);
+  } else {
+    RedisModule_Log(ctx, "notice", "Scanning indexes in background");
   }
 
-  //  RedisModule_ThreadSafeContextUnlock(ctx);
-  RedisModule_ScanCursorDestroy(cursor);
+  if (!sp) {
+    Indexes_SetTempSpecsTimers();
+  }
 
-  // set the timers for temporary indexes:
-  dictIterator *iter = dictGetIterator(specDict);
-  dictEntry *entry = NULL;
-  while ((entry = dictNext(iter))) {
-    IndexSpec *sp = dictGetVal(entry);
-    if (!(sp->flags & Index_Temporary)) {
+  while (RedisModule_Scan(ctx, cursor, Indexes_ScanCallback, scanner)) {
```

Comment:
RM_Scan might (and probably) scan more than one key in a single raw.

Reply:
Also you want to release to GIL in between calls to RM_Scan to not block the Redis during the entire scan.